### ISSUE TITLE
Option to produce output sorted by "biggest offenders"

### ISSUE THREAD

**Issue by ifreecarve** [bbatsov/rubocop#1931]
Imagine using rubocop on an existing project that has (literally) over 1000 style issues.  Where would you start?  It helps to be able to group the issues by which cop is catching them.  

You can do this via the shell as follows:
```
$ rubocop -D | awk -F":" '{print $5}' | grep -v '^$' | sort | uniq -c | sort

               [ ... contents snipped ... ]

   7  Metrics/PerceivedComplexity
   7  Style/EmptyLiteral
   7  Style/HashSyntax
   7  Style/Lambda
   8  Metrics/CyclomaticComplexity
   9  Style/CaseIndentation
   9  Style/IfUnlessModifier
   9  Style/TrailingComma
  10  Lint/UnusedMethodArgument
  10  Style/SpaceAfterMethodName
  13  Style/EmptyLinesAroundMethodBody
  14  Style/AccessorMethodName
  15  Style/AndOr
  17  Style/EmptyLinesAroundModuleBody
  19  Style/Not
  20  Style/RedundantReturn
  21  Style/SpaceInsideBlockBraces
  24  Style/FileName
  25  Style/Documentation
  30  Metrics/AbcSize
  38  Metrics/MethodLength
  43  Style/EmptyLines
  44  Style/EmptyLinesAroundClassBody
  45  Style/MethodDefParentheses
 257  Style/StringLiterals
```

Of course, this is somewhat of a hack -- it assumes that any line with 4 `:` characters is an error line.  It seems like a reasonable feature for rubocop to include a command line option to support this type of output.

**Comment by bquorning** [comment]
I think you’re looking for `rubocop --format offenses`.

**Comment by ifreecarve** [comment]
Wow, exactly.  I actually looked around for this for a bit, but I skipped over the formatters section after reading the first 5 or so -- they appeared to be just different ways of representing the same data.  

One thing that would have helped me immediately is to make the command line help more descriptive for this formatter.  It's not immediately clear that one report is not like the others, and so I dismissed "formatters" as somewhere I should look for more help on the report I wanted.

    -f, --format FORMATTER           Choose an output formatter. This option
                                     can be specified multiple times to enable
                                     multiple formatters at the same time.
                                       [p]rogress (default)
                                       [s]imple
        [ snip ] 
                                       [o]ffenses


I'd have it say something like `[o]ffenses (grouped by frequency)`.

**Comment by bquorning** [comment]
I agree, the formatters aren’t described very well in the help section. They are however thoroughly documented [in the README](https://github.com/bbatsov/rubocop/blob/master/README.md#offense-count-formatter).

Feel free to open a PR – the files are lib/rubocop/options.rb and README.md.

**Comment by ifreecarve** [comment]
I wouldn't mind opening a PR, but I feel like the proper thing to do would be refactor a little bit so that you have a `--report-type`, with options like "show-cops", "offenses grouped by frequency", and "offense list" (etc).  `--format` would then be a sub-option for the "offense list" report type (even though the formatters do not have that limitation).  

You probably have good reason(s) not to go in that direction, but I'm curious what they might be.  Is this a good idea?

**Comment by bquorning** [comment]
I think that’s a question for @bbatsov or @jonas054.

**Comment by jonas054** [comment]
Dividing the current `--formatter` option into two options that must be combined to produce the desired output makes the whole thing more difficult to understand and use IMO. But adding the `(grouped by frequency)` explanation in the help text would probably by an improvement.

> Imagine using rubocop on an existing project that has (literally) over 1000 style issues. Where would you start?

My answer to this question is that you should run `rubocop --auto-gen-config`.

**Comment by ifreecarve** [comment]
That's true.  But do you see my point that rubocop has many many options, most of which are relevant only to the act of individually listing all the problems found in your code.

`--auto-gen-config`, `--show-cops`, and the "offenses" format are the only exceptions to this rule.  To me (relatively new user, admittedly), it makes sense that these would all be enumerated in the help file and the command line help as the special cases that they are.

**Comment by jonas054** [comment]
To me, `offenses` has a lot more in common with the other formatters than with the "exceptions". It presents a report based on the inspection of files. It just happens to leave out the file names. The `files` formatter presents only the file names, leaving out the offenses, and `disabled cops via inline comments` also has a limited output compared to most formatters. But all these are formatters and should be shown together in the README and in the help text.

I can see that options with a similar purpose would benefit from being listed closer together. We have options that deal with
* output (including formatters)
* cop selection
* how to exit
* other
